src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -0,0 +1,749 @@
+            f.name -> unionType(child.typ.rvRowType.field(MatrixType.entriesIdentifier).typ,
+              f.typ,
+              dep.rvRowType.field(MatrixType.entriesIdentifier).typ)
+          else
+            f.name -> f.typ
+        }: _*), rowKey = child.typ.rowKey)
+        val child2 = rebuild(child, unionBaseType(child.typ, depMod, irDep))
+        MatrixMapRows(child2, rebuildIR(newRow, child2.typ, memo), newKey)
+      case x@MatrixMapCols(child, newCol, newKey) =>
+        val (irDep, memo) = getDeps(newCol, dep.colType, child.typ)
+        val child2 = rebuild(child, unionBaseType(child.typ, dep.copy(
+          colType = irDep.colType,
+          colKey = child.typ.colKey
+        ), irDep))
+        MatrixMapCols(child2, rebuildIR(newCol, child2.typ, memo), newKey)
+      case x@MatrixMapGlobals(child, newRow, value) =>
+        val (irDep, memo) = getDeps(newRow, dep.globalType, child.typ)
+        // fixme push down into value
+        val child2 = rebuild(child, unionBaseType(child.typ, dep.copy(globalType = irDep.globalType), irDep))
+        MatrixMapGlobals(child2, rebuildIR(newRow, child2.typ, memo, "value" -> value.t), value)
+      case x@MatrixRead(_, _, _, _, _) => upcast(x, dep)
+      case x@MatrixLiteral(typ, value) => x
+      case x@FilterCols(child, cond) =>
+        FilterCols(rebuild(child, child.typ), cond)
+      case x@FilterRows(child, cond) =>
+        FilterRows(rebuild(child, child.typ), cond)
+      case x@ChooseCols(child, oldIndices) =>
+        ChooseCols(rebuild(child, dep), oldIndices)
+      case x@CollectColsByKey(child) =>
+        val colKeySet = dep.colKey.toSet
+        val explodedDep = dep.copy(
+          colType = TStruct(dep.colType.required, dep.colType.fields.map { f =>
+            if (colKeySet.contains(f.name))
+              f.name -> f.typ
+            else {
+              f.name -> f.typ.asInstanceOf[TArray].elementType
+            }
+          }: _*),
+          rvRowType = dep.rvRowType.copy(fields = dep.rvRowType.fields.map { f =>
+            if (f.name == MatrixType.entriesIdentifier)
+              f.copy(typ = TArray(
+                TStruct(dep.entryType.required, dep.entryType.fields.map(ef =>
+                  ef.name -> ef.typ.asInstanceOf[TArray].elementType): _*), f.typ.required))
+            else
+              f
+          })
+        )
+        CollectColsByKey(rebuild(child, explodedDep))
+    }
+  }
+
+  def rebuild(ir: IR): IR = {
+    ir match {
+      case TableCount(child) => TableCount(rebuild(child, minimal(child.typ)))
+      case TableWrite(child, path, overwrite, spec) => TableWrite(rebuild(child, child.typ), path, overwrite, spec)
+      case TableExport(child, path, typesFile, header, exportType) =>
+        TableExport(rebuild(child, child.typ), path, typesFile, header, exportType)
+      case TableAggregate(child, query) =>
+        val (queryDep, memo) = getDeps(query, query.typ, child.typ)
```

Comment:
I'm slightly confused why rebuild is calling getDeps and not looking at memo.  I'm worried about the the complexity if you need to reanalyze at each point, that would make the method quadratic.

Reply:
let's discuss in person today.